### PR TITLE
Backport of amgetbitmap AM functions

### PR DESCRIPTION
**Pull request by danielgustafsson**
This backports the upstream commits which replaces the amgetmulti AM
function with amgetbitmap, which performs the whole indexscan in one
call (for HashBitmap, StreamBitmaps are not affected by this). GPDB
was more or less already doing this, as the upstream patch was from
the beginning submitted from Greenplum. This commit refactors the
AM function to mimick the upstream behavior, while keeping the GPDB
API for the callsites.

The below commits are included either in full, or in part:

  commit 4e82a954764af0b03c54cd4db422a6b4e585e4e8
  Author: Tom Lane <tgl@sss.pgh.pa.us>
  Date:   Thu Apr 10 22:25:26 2008 +0000

    Replace "amgetmulti" AM functions with "amgetbitmap", in which the whole
    indexscan always occurs in one call, and the results are returned in a
    TIDBitmap instead of a limited-size array of TIDs.  This should improve
    speed a little by reducing AM entry/exit overhead, and it is necessary
    infrastructure if we are ever to support bitmap indexes.

    In an only slightly related change, add support for TIDBitmaps to preserve
    (somewhat lossily) the knowledge that particular TIDs reported by an index
    need to have their quals rechecked when the heap is visited.  This facility
    is not really used yet; we'll need to extend the forced-recheck feature to
    plain indexscans before it's useful, and that hasn't been coded yet.
    The intent is to use it to clean up 8.3's horrid @@@ kluge for text search
    with weighted queries.  There might be other uses in future, but that one
    alone is sufficient reason.

    Heikki Linnakangas, with some adjustments by me.

  commit 1dcf6fdf1bd7fb2501445fa2ebfa57e04273e848
  Author: Teodor Sigaev <teodor@sigaev.ru>
  Date:   Sat Aug 23 10:37:24 2008 +0000

    Fix possible duplicate tuples while  GiST scan. Now page is processed
    at once and ItemPointers are collected in memory.

    Remove tuple's killing by killtuple() if tuple was moved to another
    page - it could produce unaceptable overhead.

    Backpatch up to 8.1 because the bug was introduced by GiST's concurrency support.

  commit b9856b67a7a762ab5bedfafbda4f654797d83996
  Author: Teodor Sigaev <teodor@sigaev.ru>
  Date:   Wed Oct 22 12:53:56 2008 +0000

    Fix GiST's killing tuple: GISTScanOpaque->curpos wasn't
    correctly set. As result, killtuple() marks as dead
    wrong tuple on page. Bug was introduced by me while fixing
    possible duplicates during GiST index scan.

### CODE REVIEW

File: src/include/catalog/pg_am.h [hunk 1]
```diff
@@ -124,16 +124,16 @@ GPDB_COLUMN_DEFAULT(pg_am_amcanshrink, t);
...
 DESCR("hash index access method");
 #define HASH_AM_OID 405
-DATA(insert OID = 783 (  gist	0 7 f f t t t t t t gistinsert gistbeginscan gistgettuple gistgetmulti gistrescan gistendscan gistmarkpos gistrestrpos gistbuild gistbulkdelete gistvacuumcleanup gistcostestimate gistoptions ));
+DATA(insert OID = 783 (  gist	0 7 f f t t t t t t gistinsert gistbeginscan gistgettuple gistgetbitmap gistrescan gistendscan gistmarkpos gistrestrpos gistbuild gistbulkdelete gistvacuumcleanup gistcostestimate gistoptions ));
 DESCR("GiST index access method");
 #define GIST_AM_OID 783
-DATA(insert OID = 2742 (  gin	0 4 f f f f f f t f gininsert ginbeginscan gingettuple gingetmulti ginrescan ginendscan ginmarkpos ginrestrpos ginbuild ginbulkdelete ginvacuumcleanup gincostestimate ginoptions ));
+DATA(insert OID = 2742 (  gin	0 4 f f f f f f t f gininsert ginbeginscan gingettuple gingetbitmap ginrescan ginendscan ginmarkpos ginrestrpos ginbuild ginbulkdelete ginvacuumcleanup gincostestimate ginoptions ));
 DESCR("GIN index access method");
 #define GIN_AM_OID 2742
```
Comment (hlinnaka) [inline src/include/catalog/pg_am.h:139]: Let's also rename the bitmap index function `bmgetmulti` to `bmgetbitmap`

**Comment by hlinnaka** [comment]
LGTM, after renaming `bmgetmulti` to `bmgetbitmap`. I presume we'll re-create this PR against the master repository, as soon as `iteration_2` branch has been merged?

**Comment by danielgustafsson** [comment]
> @hlinnaka commented on this pull request.
> 
> In src/include/catalog/pg_am.h:
> 
> >  DESCR("GiST index access method");
>  #define GIST_AM_OID 783
> -DATA(insert OID = 2742 (  gin	0 4 f f f f f f t f gininsert ginbeginscan gingettuple gingetmulti ginrescan ginendscan ginmarkpos ginrestrpos ginbuild ginbulkdelete ginvacuumcleanup gincostestimate ginoptions ));
> +DATA(insert OID = 2742 (  gin	0 4 f f f f f f t f gininsert ginbeginscan gingettuple gingetbitmap ginrescan ginendscan ginmarkpos ginrestrpos ginbuild ginbulkdelete ginvacuumcleanup gincostestimate ginoptions ));
>  DESCR("GIN index access method");
>  #define GIN_AM_OID 2742
>  
> Let's also rename the bitmap index function bmgetmulti to bmgetbitmap

Doh, of course. Was paying so close attention to the upstream commit I totally missed that one. Fixing.

**Comment by danielgustafsson** [comment]
> On 21 Sep 2017, at 11:12, Heikki Linnakangas <notifications@github.com> wrote:
> 
> LGTM, after renaming bmgetmulti to bmgetbitmap. I presume we'll re-create this PR against the master repository, as soon as iteration_2 branch has been merged?

I think that makes sense, the other alternative is to just use this diff in the merge to reduce conflicts. The former has the advantage of clarity though since this isn’t just a straight conflict resolution but an intentional code change in GPDB.

**Comment by hlinnaka** [comment]
On 09/21/2017 12:32 PM, Daniel Gustafsson wrote:
>> On 21 Sep 2017, at 11:12, Heikki Linnakangas
>> <notifications@github.com> wrote:
>> 
>> LGTM, after renaming bmgetmulti to bmgetbitmap. I presume we'll
>> re-create this PR against the master repository, as soon as
>> iteration_2 branch has been merged?
> 
> I think that makes sense, the other alternative is to just use this
> diff in the merge to reduce conflicts. The former has the advantage
> of clarity though since this isn’t just a straight conflict
> resolution but an intentional code change in GPDB.

Yeah, especially if we also do the other changes we discussed, adding an 
extra argument to the function that amgetbitmap can use to return the 
StreamBitmap, and using an upstream-like HashBitmap otherwise.

- Heikki

**Comment by danielgustafsson** [comment]
This is long since done, but forgot to close this so doing that now.

**Pull request closed without merging**